File: frontend/src/api/onboarding.ts
/*
 * Copyright (c) 2025 Laurent Barbe
 * Licensed under the Apache License, Version 2.0
 */
import client, { LONG_RUNNING_REQUEST_TIMEOUT_MS } from "./client";

export type OnboardingStatus = {
  dismissed: boolean;
  complete: boolean;
  endpoint_configured: boolean;
  storage_access_configured: boolean;
  source?: "quickstart" | "standard";
  journeys: OnboardingJourney[];
  can_configure: boolean;
  actor_id: number;
};

export type OnboardingPreview = {
  features: string[];
  changes: string[];
  blockers: string[];
  review_token: string;
};

export type OnboardingDraft = {
  version: 2;
  endpoint_id: number | null;
  endpoint_url: string;
  region: string;
  force_path_style: boolean;
  manager: boolean;
  portal: boolean;
  private_connection: boolean;
  ceph_admin: boolean;
  supervision: boolean;
};

export type OnboardingJourney = {
  id: string;
  revision: number;
  draft: OnboardingDraft;
  resources: {
    endpoint_id?: number;
    endpoint_name?: string;
    account_id?: number;
    account_name?: string;
    connection_id?: number;
    rgw_account_id?: string;
  };
  pending_step: string | null;
  configured: boolean;
  preview: OnboardingPreview;
  links: Partial<Record<"manager" | "portal" | "browser" | "private_manager" | "ceph_admin", string>>;
  created_at: string;
  updated_at: string;
};

type OnboardingApplyCredentials = {
  admin_access_key?: string;
  admin_secret_key?: string;
  supervision_access_key?: string;
  supervision_secret_key?: string;
  ceph_admin_access_key?: string;
  ceph_admin_secret_key?: string;
  private_access_key?: string;
  private_secret_key?: string;
};

export const ONBOARDING_STATUS_EVENT = "bucketreef:onboarding-status";

export function announceOnboardingStatus(status: OnboardingStatus): void {
  if (typeof window !== "undefined") {
    window.dispatchEvent(new CustomEvent<OnboardingStatus>(ONBOARDING_STATUS_EVENT, { detail: status }));
  }
}

export async function fetchOnboardingStatus(): Promise<OnboardingStatus> {
  const { data } = await client.get<OnboardingStatus>("/admin/onboarding");
  return data;
}

export async function dismissOnboarding(): Promise<OnboardingStatus> {
  const { data } = await client.post<OnboardingStatus>("/admin/onboarding/dismiss");
  announceOnboardingStatus(data);
  return data;
}

export async function resumeOnboarding(): Promise<OnboardingStatus> {
  const { data } = await client.post<OnboardingStatus>("/admin/onboarding/resume");
  announceOnboardingStatus(data);
  return data;
}

export async function saveOnboardingJourney(
  id: string,
  draft: OnboardingDraft,
  revision?: number,
): Promise<OnboardingJourney> {
  const { data } = await client.put<OnboardingJourney>(`/admin/onboarding/journeys/${id}`, { draft, revision });
  return data;
}

export async function applyOnboardingJourney(
  journey: OnboardingJourney,
  credentials: OnboardingApplyCredentials,
): Promise<OnboardingJourney> {
  const { data } = await client.post<OnboardingJourney>(
    `/admin/onboarding/journeys/${journey.id}/apply`,
    {
      revision: journey.revision,
      confirmed: true,
      review_token: journey.preview.review_token,
      ...credentials,
    },
    { timeout: LONG_RUNNING_REQUEST_TIMEOUT_MS },
  );
  return data;
}

export async function previewOnboardingDraft(
  draft: OnboardingDraft,
  signal?: AbortSignal,
  journeyId?: string,
): Promise<OnboardingPreview> {
  const { data } = await client.post<OnboardingPreview>("/admin/onboarding/preview", draft, {
    signal,
    params: { journey_id: journeyId },
  });
  return data;
}

File: backend/app/models/onboarding.py
# Copyright (c) 2025 Laurent Barbe
# Licensed under the Apache License, Version 2.0
from datetime import datetime
from typing import Literal
from urllib.parse import urlsplit

from pydantic import ConfigDict, Field, SecretStr, model_validator

from app.models.base import ApiModel


class OnboardingDraft(ApiModel):
    """Secret-free setup intent for the simplified onboarding flow."""

    model_config = ConfigDict(extra="forbid")

    version: Literal[2] = 2
    endpoint_id: int | None = Field(default=None, gt=0)
    endpoint_url: str = Field(default="", max_length=2048)
    region: str = Field(default="", max_length=100)
    force_path_style: bool = True
    manager: bool = False
    portal: bool = False
    private_connection: bool = False
    ceph_admin: bool = False
    supervision: bool = False

    @model_validator(mode="after")
    def validate_endpoint(self):
        if self.endpoint_id and self.endpoint_url:
            raise ValueError("Choose an existing endpoint or enter a new endpoint URL")
        if self.endpoint_url:
            url = urlsplit(self.endpoint_url)
            if (
                url.scheme not in {"http", "https"}
                or not url.hostname
                or url.username
                or url.password
                or url.query
                or url.fragment
            ):
                raise ValueError(
                    "Use an HTTP(S) endpoint without credentials, query or fragment"
                )
        return self

    @property
    def selected_options(self) -> tuple[str, ...]:
        return tuple(
            option
            for option in (
                "manager",
                "portal",
                "private_connection",
                "ceph_admin",
                "supervision",
            )
            if getattr(self, option)
        )


class OnboardingSave(ApiModel):
    model_config = ConfigDict(extra="forbid")
    draft: OnboardingDraft
    revision: int | None = Field(default=None, gt=0)


class OnboardingApply(ApiModel):
    model_config = ConfigDict(extra="forbid")
    revision: int = Field(gt=0)
    confirmed: Literal[True]
    review_token: str = Field(min_length=64, max_length=64)
    # Write-only credentials. They are never copied into draft/progress/audit data.
    admin_access_key: SecretStr | None = None
    admin_secret_key: SecretStr | None = None
    supervision_access_key: SecretStr | None = None
    supervision_secret_key: SecretStr | None = None
    ceph_admin_access_key: SecretStr | None = None
    ceph_admin_secret_key: SecretStr | None = None
    private_access_key: SecretStr | None = None
    private_secret_key: SecretStr | None = None


class OnboardingPreview(ApiModel):
    features: list[str] = Field(default_factory=list)
    changes: list[str] = Field(default_factory=list)
    blockers: list[str] = Field(default_factory=list)
    review_token: str = ""


class OnboardingJourneyOut(ApiModel):
    id: str
    revision: int
    draft: OnboardingDraft
    resources: dict = Field(default_factory=dict)
    pending_step: str | None = None
    configured: bool = False
    preview: OnboardingPreview
    links: dict[str, str] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class OnboardingStatus(ApiModel):
    dismissed: bool
    complete: bool
    endpoint_configured: bool
    storage_access_configured: bool
    source: Literal["quickstart", "standard"] = "standard"
    journeys: list[OnboardingJourneyOut] = Field(default_factory=list)
    can_configure: bool = False
    actor_id: int = 0
